src/blockchain/fork_choice.cpp

```diff
@@ -149,125 +153,273 @@ namespace lean {
+          continue;
+        }
+
+        if (not std::ranges::contains(attestations,
+                                      signed_attestation.message)) {
+          new_attestations = true;
+          attestations.push_back(signed_attestation.message);
+          signatures.push_back(signed_attestation.signature);
+        }
+      }
+
+      // Fixed point reached: no new attestations found
+      if (not new_attestations) {
+        break;
+      }
+    }
+
+    // Create final block with all collected attestations
     Block block{
         .slot = slot,
         .proposer_index = validator_index,
         .parent_root = head_root,
-        .state_root = {},  // to be filled after state transition
+        // Will be updated with computed hash
+        .state_root = {},
+        .body = {.attestations = attestations},
     };
-    for (auto &signed_vote : latest_known_votes_ | std::views::values) {
-      block.body.attestations.push_back(signed_vote);
-    }
+    // Apply state transition to get final post-state and compute state root
     BOOST_OUTCOME_TRY(auto state,
                       stf_.stateTransition(block, head_state, false));
     block.state_root = sszHash(state);
     block.setHash();
 
-    BOOST_OUTCOME_TRY(onBlock(block));
-
-    return block;
+    auto proposer_attestation = produceAttestation(slot, validator_index);
+    proposer_attestation.data.head = Checkpoint::from(block);
+
+    // Store block and state in forkchoice store
+    auto signed_block_with_attestation = signBlock({
+        .message =
+            {
+                .block = block,
+                .proposer_attestation = proposer_attestation,
+            },
+        .signature = signatures,
+    });
+    BOOST_OUTCOME_TRY(onBlock(signed_block_with_attestation));
+
+    return signed_block_with_attestation;
   }
 
+  Attestation ForkChoiceStore::produceAttestation(
+      Slot slot, ValidatorIndex validator_index) {
+    // Get the head block the validator sees for this slot
+    Checkpoint head_checkpoint{
+        .root = head_,
+        .slot = getHeadSlot(),
+    };
+
+    //  Calculate the target checkpoint for this attestation
+    //
+    //  This uses the store's current forkchoice state to determine
+    //  the appropriate attestation target, balancing between head
+    //  advancement and safety guarantees.
+    auto target_checkpoint = getAttestationTarget();
+
+    // Construct attestation data
+    AttestationData attestation_data{
+        .slot = slot,
+        .head = head_checkpoint,
+        .target = target_checkpoint,
+        .source = latest_justified_,
+    };
+
+    return Attestation{
+        .validator_id = validator_index,
+        .data = attestation_data,
+    };
+  }
 
   outcome::result<void> ForkChoiceStore::validateAttestation(
-      const SignedVote &signed_vote) {
+      const SignedAttestation &signed_attestation) {
+    auto &attestation = signed_attestation.message;
+    auto &data = attestation.data;
```

Comment:
Minor
```suggestion
    auto &target = signed_attestation.message.data.target;
    auto &source = signed_attestation.message.data.source;
```
It may be more readable